File: src/cli/commands/mappings.ts
// src/cli/commands/mappings.ts

import { Command } from 'commander';
import chalk from 'chalk';
import { createMapping, getAllMappings, deleteMapping } from '../api_client.js';
import ora from 'ora';
import { handleCliError } from '../error_handler.js';

export function registerMappingsCommand(program: Command) {
    const mappingsCommand = new Command('mappings')
        .description('Manage component-to-test mappings');

    mappingsCommand
        .command('list')
        .description('List all existing test mappings')
        .action(async () => {
            try {
                const mappings = await getAllMappings();
                if (mappings.length === 0) {
                    console.log(chalk.yellow('No mappings found.'));
                    return;
                }
                const displayData = mappings.map(m => ({
                    'Mapping ID': m.id,
                    'Main Component ID': m.mainComponentId,
                    'Test Component ID': m.testComponentId,
                    'Test Name': m.testComponentName || 'N/A',
                }));
                console.table(displayData);
            } catch (error: any) {
                handleCliError(error);
            }
        });

    mappingsCommand
        .command('add')
        .description('Add a new test mapping')
        .requiredOption('--mainId <id>', 'The main component ID')
        .requiredOption('--testId <id>', 'The test component ID')
        .option('--name <name>', 'An optional descriptive name for the test')
        .action(async (options) => {
            const spinner = ora('Adding new mapping...').start();
            try {
                const newMapping = await createMapping({
                    mainComponentId: options.mainId,
                    testComponentId: options.testId,
                    testComponentName: options.name,
                });
                spinner.succeed(chalk.green('Mapping created successfully!'));
                console.log(chalk.cyan(`Mapping ID: ${newMapping.id}`));
            } catch (error: any) {
                spinner.fail(chalk.red('Failed to create mapping.'));
                handleCliError(error);
            }
        });

    mappingsCommand
        .command('rm')
        .description('Remove a test mapping by its unique ID')
        .argument('<mappingId>', 'The unique ID of the mapping to remove')
        .action(async (mappingId) => {
            const spinner = ora(`Removing mapping ${mappingId}...`).start();
            try {
                await deleteMapping(mappingId);
                spinner.succeed(chalk.green('Mapping removed successfully!'));
            } catch (error: any) {
                spinner.fail(chalk.red('Failed to remove mapping.'));
                handleCliError(error);
            }
        });

    program.addCommand(mappingsCommand);
}

File: automated-test-orchestrator-cli/cmd/mappings.go
// cli-go/cmd/mappings.go
package cmd

import (
	"fmt"
	"os"
	"time"

	"github.com/automated-test-orchestrator/cli-go/internal/client"
	"github.com/automated-test-orchestrator/cli-go/internal/csv"
	"github.com/automated-test-orchestrator/cli-go/internal/display"
	"github.com/automated-test-orchestrator/cli-go/internal/errors"
	"github.com/automated-test-orchestrator/cli-go/internal/model"
	"github.com/briandowns/spinner"
	"github.com/fatih/color"
	"github.com/spf13/cobra"
	"github.com/spf13/viper"
)

// mappingsCmd represents the mappings command group.
var mappingsCmd = &cobra.Command{
	Use:   "mappings",
	Short: "Manage component-to-test mappings",
}

// mappingsListCmd represents the 'mappings list' command.
var mappingsListCmd = &cobra.Command{
	Use:   "list",
	Short: "List all existing test mappings",
	Run: func(cmd *cobra.Command, args []string) {
		apiClient := client.NewAPIClient(viper.GetString("api_url"))
		mappings, err := apiClient.GetAllMappings()
		if err != nil {
			errors.HandleCLIError(nil, err)
		}

		if len(mappings) == 0 {
			color.Yellow("No mappings found.")
			return
		}

		display.PrintMappings(mappings)
	},
}

// mappingsAddCmd represents the 'mappings add' command.
var mappingsAddCmd = &cobra.Command{
	Use:   "add",
	Short: "Add a new test mapping",
	Run: func(cmd *cobra.Command, args []string) {
		s := spinner.New(spinner.CharSets[11], 100*time.Millisecond)
		s.Suffix = " Adding new mapping..."
		s.Start()

		apiClient := client.NewAPIClient(viper.GetString("api_url"))
		mainID, _ := cmd.Flags().GetString("mainId")
		testID, _ := cmd.Flags().GetString("testId")
		name, _ := cmd.Flags().GetString("name")

		req := model.CreateMappingRequest{
			MainComponentID: mainID,
			TestComponentID: testID,
		}
		if name != "" {
			req.TestComponentName = &name
		}

		newMapping, err := apiClient.CreateMapping(req)
		if err != nil {
			errors.HandleCLIError(s, err)
		}

		s.Stop()
		color.Green("✅ Mapping created successfully!")
		fmt.Printf("Mapping ID: %s\n", color.CyanString(newMapping.ID))
	},
}

// mappingsImportCmd represents the 'mappings import' command.
var mappingsImportCmd = &cobra.Command{
	Use:   "import",
	Short: "Bulk import mappings from a CSV file",
	Run: func(cmd *cobra.Command, args []string) {
		s := spinner.New(spinner.CharSets[11], 100*time.Millisecond)
		s.Suffix = " Preparing to import mappings..."
		s.Start()

		apiClient := client.NewAPIClient(viper.GetString("api_url"))
		csvPath, _ := cmd.Flags().GetString("from-csv")

		file, err := os.Open(csvPath)
		if err != nil {
			errors.HandleCLIError(s, fmt.Errorf("failed to open file %s: %w", csvPath, err))
		}
		defer file.Close()

		mappingsToCreate, err := csv.ParseMappingCsv(file)
		if err != nil {
			errors.HandleCLIError(s, fmt.Errorf("failed to parse CSV file: %w", err))
		}

		if len(mappingsToCreate) == 0 {
			s.Stop()
			color.Yellow("No valid mappings found in the CSV file.")
			return
		}
		s.Suffix = fmt.Sprintf(" Found %d mappings to import.", len(mappingsToCreate))
		time.Sleep(1 * time.Second) // Pause to show message

		var successCount, failureCount int
		for i, mapping := range mappingsToCreate {
			s.Suffix = fmt.Sprintf(" Importing mapping %d of %d: %s -> %s", i+1, len(mappingsToCreate), mapping.MainComponentID, mapping.TestComponentID)
			_, err := apiClient.CreateMapping(mapping)
			if err != nil {
				failureCount++
				s.Stop()
				// Use the non-terminating formatter for the loop.
				errorMsg := errors.FormatError(err)
				fmt.Fprintf(os.Stderr, "\n%s Failed to import row %d: %s\n", color.RedString("✗"), i+2, errorMsg)
				s.Start() // Restart spinner for the next item
			} else {
				successCount++
			}
		}

		s.Stop()
		fmt.Println(color.GreenString("\n--- Import Complete ---"))
		if successCount > 0 {
			color.Green("Successfully imported %d mapping(s).", successCount)
		}
		if failureCount > 0 {
			color.Red("Failed to import %d mapping(s). See error details above.", failureCount)
		}
	},
}

// mappingsRmCmd represents the 'mappings rm' command.
var mappingsRmCmd = &cobra.Command{
	Use:   "rm <mappingId>",
	Short: "Remove a test mapping by its unique ID",
	Args:  cobra.ExactArgs(1),
	Run: func(cmd *cobra.Command, args []string) {
		mappingID := args[0]
		s := spinner.New(spinner.CharSets[11], 100*time.Millisecond)
		s.Suffix = fmt.Sprintf(" Removing mapping %s...", mappingID)
		s.Start()

		apiClient := client.NewAPIClient(viper.GetString("api_url"))
		if err := apiClient.DeleteMapping(mappingID); err != nil {
			errors.HandleCLIError(s, err)
		}

		s.Stop()
		color.Green("✅ Mapping removed successfully!")
	},
}

func init() {
	rootCmd.AddCommand(mappingsCmd)

	// List command
	mappingsCmd.AddCommand(mappingsListCmd)

	// Add command with flags
	mappingsCmd.AddCommand(mappingsAddCmd)
	mappingsAddCmd.Flags().String("mainId", "", "The main component ID (required)")
	mappingsAddCmd.Flags().String("testId", "", "The test component ID (required)")
	mappingsAddCmd.Flags().String("name", "", "An optional descriptive name for the test")
	mappingsAddCmd.MarkFlagRequired("mainId")
	mappingsAddCmd.MarkFlagRequired("testId")

	// Import command with flags
	mappingsCmd.AddCommand(mappingsImportCmd)
	mappingsImportCmd.Flags().String("from-csv", "", "Path to a CSV file for bulk import (required)")
	mappingsImportCmd.MarkFlagRequired("from-csv")

	// Remove command
	mappingsCmd.AddCommand(mappingsRmCmd)
}
